[1단계 - 자동차 경주 구현] 켈리(양성욱) 미션 제출합니다.

kelly6bf opened the pull request.

kelly6bf committed:
docs: 기능 명세서 초안 작성

kelly6bf committed:
feat: 경주에 참여할 자동차 이름 목록 입력 구현

경주에 참여할 자동차 이름 목록을 입력받는 기능을 구현함.
성공 케이스만 구현되어 있기 때문에 입력값 검증 및 재입력 기능 추가 필요.

kelly6bf committed:
docs: 기능 명세서 제목 수정 및 구현 현황 제거

기능 명세서 제목을 수정함.
구현 현황은 불필요하다고 판단되어 문서에서 제거함.

kelly6bf committed:
feat: 경주를 시도할 횟수 입력 구현

kelly6bf committed:
feat: 난수 생성 기능 구현

입력된 범위 내 정수를 추출해서 반환하는 기능을 구현함.

kelly6bf committed:
feat: 자동차 이동 기능 구현

kelly6bf committed:
feat: 자동차 레이싱 기능 구현

자동차 객체들을 초기화 하는 CarFactory를 생성함.
자동차 객체들을 관리하는 일급 객체 Cars를 정의함.
자동차 상태를 표현하기 위한 CarStatus를 정의하고 Car 객체에 관련 메서드를 추가함.
변경된 로직에 맞춰 테스트를 수정함.

kelly6bf committed:
feat: 실시간 경주 상황을 출력하는 기능 구현

kelly6bf committed:
refactor: 레이싱 결과를 출력하는 문구의 제목을 별도 메서드에서 출력하도록 분리

kelly6bf committed:
feat: 레이싱 우승자 판단 기능 구현

kelly6bf committed:
feat: 우승자 출력 기능 구현

kelly6bf committed:
feat: 자동차 경주 컨트롤러 클래스 작성

도메인간 행위를 연결하는 Controller 클래스를 작성함.

kelly6bf committed:
feat: RacingCarApplication 클래스 작성

AppCofig를 통해 Application에 필요한 의존성을 설정함.

kelly6bf committed:
docs: 기능 명세서 기능 이름 수정

kelly6bf committed:
refactor: InputView 입력 안내 메시지 추가 및 패키지 경로 변경

kelly6bf committed:
docs: 기능 명세서 예외케이스 추가

kelly6bf committed:
refactor: 자동차 이름 검증 코드 추가

자동차 이름을 검증하는 코드를 추가하고 테스트를 수정함.

kelly6bf committed:
docs:  자동차 수 입력값 검증 조건 변경

kelly6bf committed:
refactor: 자동차 이름 입력값 개수 검증 코드 추가

kelly6bf committed:
refactor: CarFactory 중복된 자동차 이름 검증 로직 추가

kelly6bf committed:
reafactor: 입력된 자동차 이름 문자열에 앞뒤 공백을 제거하는 로직 추가

kelly6bf committed:
refactor: 정수 입력값 검증 로직 추가

kelly6bf committed:
fix: 입력 숫자 검증 조건 정정

kelly6bf committed:
fix: 중복된 이름 검증 조건 정정

kelly6bf committed:
fix: 숫자 입력 에러 메시지 정정

kelly6bf committed:
feat: 잘못된 값 입력시 재입력 기능 추가

kelly6bf committed:
refactor: 도메인 구조 세분화

각 도메인 클래스들을 별도의 패키지를 생성하여 이동시킴.

kelly6bf committed:
refactor: Controller -> RacingCarController 클래스 이름 변경

보다 명확한 의미를 표현하기 위해 클래스 이름을 변경함.
의존 클래스들에 변경 사항 적용.

kelly6bf committed:
refactor: view 도메인 클래스 리펙토링

모든 파라미터에 final 키워드를 붙여줌.
InputView의 inputCars 메서드 네이밍을 보다 명확하게 하기 위해 inputCarNames로 변경함.

kelly6bf committed:
refactor: 모든 도메인 클래스 파라미터에 final 키워드 추가

코드 스타일 통일을 위해 모든 도메인 클래스 파라미터에 final 키워드를 추가함.

kelly6bf committed:
refactor: RacingCarController 모든 파라미터에 final 키워드 추가 및 메서드 이름 변경 적용

kelly6bf committed:
docs: 기능 명세서 기능 요구사항 추가

경주에 참여할 자동차 이름을 입력시 중복된 값이 들어오지 못하는 기능 요구사항 추가.
경주 진행 후 모든 자동차의 최종 전진거리가 0일시 우승자를 추출하지 않는 기능 요구사항 추가.

kelly6bf committed:
style: CarFactory 불필요한 import 제거

kelly6bf committed:
test: CarFactoryTest @DisplayName 내용 수정

kelly6bf committed:
test: CarFactoryTest Cars생성 성공 케이스 테스트 추가

kelly6bf committed:
refactor: Car 생성시 난수 생성기가 입력되지 않은 경우를 검증하는 로직 추가

로직 추가로 영향을 받은 클래스의 코드를 정정해줌.

kelly6bf committed:
refactor: 도메인 클래스 예외 처리 변경

기존 RuntimeException을 발생시키던 로직들을 예외 성격을 고려하여 IllegalArgumentException을 발생시키도록 로직을 수정함.

kelly6bf committed:
refactor: 모든 자동차들의 최종 전진 거리가 0일경우 우승자를 추출하지 않도록 로직 수정

kelly6bf committed:
test: 모든 테스트 코드의 Assertions class를 static import 처리

kelly6bf committed:
docs: 기능 명세서 우승자 출력 항목 기능 요구사항 추가

Hyeon9mak commented on src/docs/README.md:
```diff
@@ -0,0 +1,24 @@
+
+## 기능 목록
+
+- 경주에 참여할 자동차 이름 목록 입력
+    - `,`로 자동차 구분
+    - 이름 길이는 1이상 5이하
+    - 중복된 자동차 이름 허용하지 않음
+    - 자동차 개수는 `2`개 이상 입력
+    - 잘못된 값을 입력받을시 에러 메시지를 출력하고 재입력
+- 경주를 시도할 횟수 입력
+    - 횟수는 `1`이상의 정수
+    - 잘못된 값을 입력받을시 에러 메시지를 출력하고 재입력
+- 자동차 이동 기능
+    - `0 ~ 9` 사이에서 랜던값을 구한다.
+    - `4`이상일 경우 전진
+    - `3`이하일 경우 멈춤
+- 우승자 판단 기능
+    - 전진 거리가 가장 높은 자동차를 선정
+    - 우승자는 여러명이 될 수 있음
+    - 자동차들의 전진거리가 모두 0일시 우승자 X
+- 경기 결과 출력
+- 우승자 출력
+    - 우승자가 없을 경우 우승자가 존재하지 않다는 내용의 메시지 출력
```
구현해야할 기능 목록을 잘 정리해주셨네요.

```
- [ ] 마크다운 문서 작성시
- [x] 이런 걸 사용해보면
```

- [ ] 이렇게
- [x] 나타납니다.

kelly6bf replied:
감사합니다 현구막! 처음에는 작성해주신 체크박스 형태로 요구사항을 작성했는데 체크 박스를 체크하는 경우는 언제인지를 페어와 논의하다 쉽게 결론은 나지 않아서 우선 - 형태로 작성했습니다! 다음 미션에서는 체크 박스를 적극 사용해보겠습니다! 😋

Hyeon9mak commented on src/main/java/config/AppConfig.java:
```diff
@@ -0,0 +1,20 @@
+import view.InputView;
+import view.OutputView;
+
+public class AppConfig {
+
+    public static RacingCarController config() {
+        return new RacingCarController(inputView(), outputView());
+    }
+
+    private static InputView inputView() {
+        return new InputView();
+    }
+
+    private static OutputView outputView() {
+        return new OutputView();
+    }
+}
```
controller 역시 view 들과 마찬가지로 static 한 동작만을 갖고 있는 것 같은데요, 별개로 인스턴스 생성 클래스를 준비하신 이유가 있을까요?

kelly6bf replied:
AppConifg와 같이 의존성을 설정해주는 클래스가 존재하면 각 클래스간 의존성 관계를 빠르게 파악할 수 있고, '의존성 주입'이라는 역할도 분리할 수 있다는 생각에 AppConfig를 작성했습니다! 하지만 지금같이 작은 프로그램에서는 조금은 오버 엔지니어링이라는 생각도 드는거 같습니다! 우선은 컨트롤러에서 View 클래스들을 생성하도록 코드를 변경하고 향 후 의존성 관계가 복잡해질거 같을때 지금의 방식을 도입해도 괜찮을 거 같다는 생각이 드네요!

Hyeon9mak replied:
`AppConfig` 를 통한 의존성 주입의도는 이해되었으나, controller 가 매번 인스턴스로 재할당 되어야 하는 이유는 아직 이해하지 못했어요 😢

Hyeon9mak commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final Scanner sc = new Scanner(System.in);
+
+    public List<String> inputCarNames() {
+        System.out.println("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).");
+        String input = sc.nextLine();
+
+        return Stream.of(input.split(","))
+                .map(String::trim)
+                .toList();
+    }
```
👍 
입력을 유도하고, 문자열을 간단하게 파싱하는 것 까진 view 의 책임으로 볼 수 있죠.
추후 "구분 기준을 `,` 대신 `#` 으로 바꿔주세요." 라는 요구사항이 전달되었을 때도 `view` 패키지만 열어보면 되겠네요.

kelly6bf replied:
감사합니다! ☺️

Hyeon9mak commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,36 @@
+                .toList();
+    }
+
+    public RacingCount inputRacingCount() {
+        System.out.println("시도할 회수는 몇회인가요?");
+        int input = parseNumber(sc.nextLine());
+
+        return new RacingCount(input);
+    }
```
view 가 `RacingCount` 를 아는 것에 대해서는 어떻게 생각하실까요? 켈리의 의견이 궁금해요.

kelly6bf replied:
이 부분은 실수한부분입니다! 페어랑도 MVC 패턴에 대해 얘기하면서 View가 도메이 객체를 알면 안된다는 얘기를 나누며 코딩했는데 시간에 쫓기다 보니 실수했네요..!

추가로 저희가 'View는 Model을 알면 안된다'라고 생각하는 이유는 관심사의 분리를 위해 사용하는 MVC 패턴의 장점이 깨질 수 있기 때문이라고 생각합니다.

Hyeon9mak commented on src/main/java/dto/CarStatus.java:
```diff
@@ -0,0 +1,4 @@
+package dto;
+
+public record CarStatus(String name, int distance) {
+}
```
record 활용 👍

kelly6bf replied:
감사합니다! ☺️

Hyeon9mak commented on src/main/java/controller/RacingCarController.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    private List<String> racing(final Cars cars, final RacingCount racingCount) {
+        outputView.printResultMessageTitle();
+        for (int i = 0; i < racingCount.getCount(); i++) {
+            List<CarStatus> raceResult = cars.race();
+            outputView.printRacingResult(raceResult);
+        }
+
+        return getWinners(cars);
+    }
```
`RacingCount` 에서 값을 꺼내어 사용하기보다, `RacingCount` 에게 게임을 종료할 만큼 진행되었는지 물어보는 건 어떨까요?

> 객체에 메세지를 보내라.

그리고 이 방식이 갖는 장점은 무엇일까요?

kelly6bf replied:
말씀하신 것 처럼 현재 구조는 RacingCount가 굉장히 수동적이라고 생각이 들어요! 

게임을 종료할지 여부를 알기 위해 필요한 데이터를 가진 객체가 그 책임을 수행하게 되면 테스트가 용이해지고 나중에 기능 요구사항이 변경되더라도 그 객체에 무언가 요청을 주는 객체에 가는 변경을 최소화 하는 이점이 있다고 생각해요.

따라서 RacingCount에 isEnd()와 같은 책임을 위임하면 더 좋은 방향으로 개선될거라고 생각이 듭니다!

Hyeon9mak commented on src/main/java/controller/RacingCarController.java:
```diff
@@ -0,0 +1,72 @@
+        return getWinners(cars);
+    }
+
+    private List<String> getWinners(final Cars cars) {
+        RacingRule racingRule = new RacingRule();
+
+        return racingRule.getWinners(cars)
+                .stream()
+                .map(car -> car.getStatus().name())
+                .toList();
+    }
```
현재 일급 컬렉션인 `Cars` 가 단순 컬렉션으로 변경되어 동작을 수행하는 걸로 보여요.
`Cars` 가 `RacingRule` 을 메서드 파라미터나, 멤버객체로 활용하여 승자들을 뽑아내는 건 어떨까요?

kelly6bf replied:
동의합니다! 일급 컬렉션에 대해 페어와 대화해보았는데 현재 구조에서는 Cars가 너무 수동적이라는 느낌도 들어요. 개선해보겠습니다!

Hyeon9mak commented on src/main/java/domain/car/Car.java:
```diff
@@ -0,0 +1,51 @@
+
+public class Car {
+
+    public static final int INITIAL_DISTANCE = 0;
```
내부에서만 사용되는걸로 보이네요!
인텔리제이 상수생성 단축키를 연속 2회 입력하면, 기본적인 접근제한 레벨 변경이 가능합니다.

kelly6bf replied:
private으로 해야하는데 오타가..🙇 수정하겠습니다!

Hyeon9mak commented on src/main/java/domain/car/CarFactory.java:
```diff
@@ -0,0 +1,37 @@
+        return new Cars(cars);
+    }
+
+    private static void validateCarNamesCount(List<String> carNames) {
+        if (carNames.size() < 2) {
+            throw new IllegalArgumentException("자동차 이름의 개수는 2이상이여야 합니다.");
+        }
+    }
+
+    private static void validateDuplicatedCarNames(List<String> carNames) {
+        int distinctSize = new HashSet<>(carNames).size();
+
+        if (carNames.size() != distinctSize) {
+            throw new IllegalArgumentException("중복된 자동차 이름을 허용하지 않습니다.");
+        }
+    }
```
각각의 검증 로직을 `Car`, `Cars` 에서 진행하는 건 어떻게 생각하시나요? 켈리의 생각이 궁금해요.

kelly6bf replied:
동의합니다! 두 검증 모두 Cars에서 충분히 책임질 수 있는 행위라고 생각해요.

페어랑 이부분에 대해 더 얘기를 나눠보면서 아예 CarFactory를 제거하고 Factory의 역할을 Cars에게 위임해도 좋다는 생각이 들었어요! 당장 Factory를 따로 분리할정도로 많은 객체를 생성하거나 많은 요구 사항을 포함하지는 않아서 우선은 Cars만 있어도 충분할거 같아요!

Hyeon9mak commented on src/test/java/domain/car/CarFactoryTest.java:
```diff
@@ -0,0 +1,49 @@
+        assertThat(cars).isNotNull();
+    }
+
+    @DisplayName("2미만 크기의 문자열 리스트가 입력되면 예외가 발생한다.")
+    @Test
+    void initCarsThrowExceptionWhenListSizeLessThanOne() throws Exception {
+        // Given
+        List<String> invalidInput = List.of("test");
+
+        // When & Then
+        assertThatThrownBy(() -> CarFactory.generateCars(invalidInput))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("자동차 이름의 개수는 2이상이여야 합니다.");
+    }
```
꼼꼼한 테스트 👍

kelly6bf replied:
꼼꼼한 리뷰 🥰

Hyeon9mak reviewed (changes requested):
안녕하세요 켈리, 자동차 미션을 함께할 현구막입니다. 🙇‍♂️ 
1단계 깔끔하게 잘 구현해주셨네요!
간단한 코멘트 남겼으니 확인해주시고, 켈리의 의견을 공유해주세요.
궁금하신 점 언제든지 코멘트 남겨주세요.

kelly6bf committed:
refactor: InputView RacingCar 의존 제거

kelly6bf committed:
feat: RacingCount 경기가 끝났는지 검증하는 기능 구현

기존 코드는 Controller가 RacingCount로 부터 경기 횟수를 받아와서 판단했기 때문에 RacingCount가 수동적인 느낌이 들었다.
이를 개선하기 위해 경기가 끝났는지 판단하는 역할을 RacingCount에 위임해주었다.
RacingCount의 기능추가로 발생한 변경점을 관련 클래스에 적용해주었다.

kelly6bf committed:
refactor: 우승자 판단 기능의 객체별 관계를 재정의

기존 구조에서는 RaceRule이 Cars를 받아 Cars로 부터 자동차 리스트 데이터를 요청 후 우승자를 판단하는 방식이었다.
기존 방식보다는 오히려 우승 판단에 필요한 Car객체들을 가지고 있는 Cars에게 RaceRule을 넘겨 판단하다록 하는 관계가 더 적절하다고 판단되어 구조를 수정함.

kelly6bf committed:
fix: Car 잘못 설정된 상수 접근 제한자 정정

kelly6bf committed:
refactor: CarFactory의 역할을 Cars에 위임

기존 CarFactory의 역할을 당장 별도 클래스로 분리해서 구현하기 보다는 Cars가 담당해도 괜찮다는 판단이 들어 역할을 위임함.

kelly6bf commented:
[질문] CarFactory를 제거하고 기능을 Cars로 위임하면서 기존 Cars에서 진행하던 테스트를 수행하기 어려워져 고민하였습니다. 이때 다른 생각이 떠올랐습니다.

Cars가 기존 테스트하던 로직은 아래 두 가지 입니다.
1. 자동차들을 경주 시킨 후 상태 리스트를 반환
2. 우승한 자동차 이름 리스트 반환

하지만 두 로직 모두 Cars가 직접적으로 수행하기 보다는 내부적으로 Car 혹은 RacingRule이 핵심 로직을 수행 후 반환한 결과에서 get을 통해 데이터만 넘기는 역할을 수행하고 있습니다.

따라서 기존 작성한 Cars 테스트는 핵심 로직을 검증하기 보다는 getter를 검증한다는 느낌이 들어 제거를 결정하였습니다. (핵심 로직들은 각각 Car, RacingRule이 테스트 수행)

이 부분에 대해 리뷰어님의 의견도 듣고 싶습니다! 🤔

Hyeon9mak commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,33 @@
+                .toList();
+    }
+
+    public int inputRacingCount() {
+        System.out.println("시도할 회수는 몇회인가요?");
+
+        return parseNumber(sc.nextLine());
+    }
+
```
> 'View는 Model을 알면 안된다'라고 생각하는 이유는 관심사의 분리를 위해 사용하는 MVC 패턴의 장점이 깨질 수 있기 때문이라고 생각합니다.

view 가 model 을 알면 안되는 이유는 무엇일까요? 어떤 장점이 깨지는 걸까요?
반대로 model 이 view 를 알면 안되는 이유도 있을까요? 🤔

kelly6bf replied:
mvc를 사용하는 이유는 언급주신 제 답변과 같이 "관심사의 분리"라고 생각합니다. 이 때 view의 관심사는 사용자의 입력을 받거나 출력하는 것 입니다.

view의 관심사(이하 책임) 중 하나인 "사용자에게 경기 결과를 출력한다."를 수행하기 위해서 view에게 필요한 요건을 생각하면 출력할 자동차의 이름과 거리 정보를 생각할 수 있습니다. view에게 이 데이터를 넘겨주는 방법은 여럿이 있을 수 있고 그 중 Car model을 직접 넘겨주어 view가 getter를 사용해서 꺼내 출력하는 방법도 프로그램이 정상적으로 실행할 수 있습니다. 하지만 view가 직접적으로 car model을 알게되면 getter 메서드 외 다른 car의 메서드를 호출할 수 있는 여지를 남길 수 있습니다.

물론 위 문제는 동료들과 정책을 설정하여(view에서는 car model의 getter만 호출할 수 있다.) 막을 수 있을지도 모릅니다. 하지만 이 코드를 개발하고 정책을 만든 사람이 아닌 처음 이 코드를 만지는 개발자가 온다면? 정책이 시간이 지나며 변질된다면? 언제까지고 view에서 getter 외 다른 메서드를 호출하지 않을거라는 보장을 할 수 없습니다. 이는 곧 굳이 MVC 패턴을 사용해서 관심사를 분리하는 이점을 깨뜨리는 행위라고 생각합니다.

가장 좋은 방법은 처음부터 이런 여지를 주지 않기위해 DTO 등을 활용해서 필요한 데이터만 '조회'할 수 있도록 하는 것입니다. 그럼 이 코드를 처음보는 개발자가 오더라도 코드만 보고 충분히 'view에서는 조회만 가능하구나'를 빠르게 파악할 수 있습니다.

Hyeon9mak commented on src/main/java/domain/racing/RacingCount.java:
```diff
@@ -0,0 +1,21 @@
+        }
+    }
+
+    public boolean isCountEnd(int count) {
+        return this.count < count;
+    }
```
> 게임을 종료할지 여부를 알기 위해 필요한 데이터를 가진 객체가 그 책임을 수행하게 되면 테스트가 용이해지고 나중에 기능 요구사항이 변경되더라도 그 객체에 무언가 요청을 주는 객체에 가는 변경을 최소화 하는 이점이 있다고 생각해요.

💯

kelly6bf replied:
🤭

Hyeon9mak commented on src/main/java/controller/RacingCarController.java:
```diff
@@ -0,0 +1,72 @@
+        return getWinners(cars);
+    }
+
+    private List<String> getWinners(final Cars cars) {
+        RacingRule racingRule = new RacingRule();
+
+        return cars.findWinnerNames(racingRule);
+    }
```
`RacingRule` 이 `Cars` 에서만 사용되는 것 같네요. 그렇다면...?

kelly6bf replied:
RacingRule이 Cars외 그 어떤 객체에서도 사용되지 않는다면 Cars에 연관 관계로 설정할 수 있겠네요!! 2단게에서 이 방식으로 로직을 수정해보겠습니다!

-------

정정합니다! 당장 프로그램에 필요한 우숭 결과 판단 조건들이 복잡하지 않아서 별로의 객체로 분리보다는 Cars에 역할을 위임해도 좋을거 같아요! (당장은요..ㅎ) 이 방식으로 진행해보겠습니다!

Hyeon9mak reviewed (approved):
안녕하세요 켈리~ 피드백 잘 반영해주셨습니다.
켈리만의 기준을 갖고 리팩터링 해나가시는 모습 멋집니다.

#### `Cars` 테스트 코드

`Cars` 가 아닌 다른 객체에서 핵심로직을 수행한다면 말씀대로 기존 불필요해진 테스트 코드를 제거하고, 핵심 로직을 품고 있는 객체에 대해 테스트 하는게 100번 옳다고 생각합니다. 잘 해주신 것 같아요.
 
다만 추가로 남긴 코멘트에서 `RacingRule` 이 어딘가에서만 사용된다... 라는 코멘트를 남겼는데
이 것에 따라 테스트 코드 위치가 다시한번 바뀔 수도 있을거 같아요 😄 👍 

다음 단계 진행과 함께 고민해주세요!
